draft/XXXX-multipart-upload/README.md

```diff
@@ -0,0 +1,115 @@
+---
+RFC: XXXX
+Author: Roni Choudhury
+Status: draft
+Created: 2020-08-04
+Last-Modified: 2020-08-04
+---
+
+# RFC XXXX: Multipart Table Upload
+
+This RFC describes how large table data can be safely and efficiently uploaded
+to Multinet, using a variation of the common multipart upload strategy common to
+many web server technologies.
+
+## Motivation
+
+Multipart upload is needed for the following reasons:
+
+1. As a practical matter, our hosting platform Heroku does not allow for
+   requests that take longer than thirty seconds to complete. Some moderately
+   sized CSV files already break this threshold, and so called "big data" tables
+   most certainly will. In order to upload such data files, we will need a way
+   to upload them in chunks.
+2. A long-running one-shot upload does not give the end user much feedback on
+   what is going on. Multipart upload would allow us to provide a more detailed
+   progress meter that can improve the user experience.
+3. If a long-running upload is interrupted for any reason, a multipart upload
+   strategy allows for resuming such uploads without losing the time already
+   spent in uploading one more initial chunks of the file.
+
+## Proposals
+
+### Summary of Approach
+
+The basic idea is as follows: the client initiates a new upload session for a
+given file; this action causes the server to create a new collection in the
+database. Then the client issues one or more "chunk" uploads, each sending a
+range of bytes, up to some size limit, to the server. The server saves each
+chunk as a document in the database collection associated with this file upload.
+Finally, the client issues a "finalize" call, which marks the database
+collection as being complete.
```

Comment:
It's generally not advisable to store large files in the db, but instead to store them on disk and to just store a filepath in the db. Is this an option for us? What is the benefit of storing the files in the db? It seems like this increases the number of db transactions a lot

Reply:
We might be able to do it by storing files on the AWS instance in some structured way, but I'm not sure what interface we'd use to manage file transfers from Python. The advice to use the DB came from others at Kitware who have done this sort of thing before.

I'm not really worried about the number of DB transactions, because the vast majority of the time the upload process will use a single chunk, and in any case these extra DB transactions because of upload will (1) be linear in the size of the files being uploaded and (2) only occur a constant number of times.

Are there other reasons it's inadvisable to store files in the DB?

Reply:
RAM usage is one thing that springs to mind, but we can cross that bridge when we get there. I do think accessing the files from outside the database would be more tricky. You'd have to have a webserver that can serve static data from a directory, which is a ton more overhead. Storing in the database seems like a good solution for now.

Reply:
If we're dealing with larger files often, I think delegating this to something like AWS S3 (which has good support for multipart upload) is a good idea, since it seems like a more robust and efficient way to upload the data. It would also let us limit the amount of ram we're using when processing the file, while keeping a simple interface for reading the file. I.e., create upload, return redirect to S3, which does upload in multiple parts, and then when the finalize request is sent to our server, we reach out to the file, and stream the bytes in to process the now fully constructed file.

However, all of that being said, I don't think we need to deal with that right now, as our files are small enough for the time being. I mainly wanted to convey the idea, to consider in case we come to a point where this implementation doesn't scale well. I'd think as the file size increases, the difference in upload performance would start to be noticeable. Plus, some of this stuff would need to be laid out for the S3 path anyway, so if we do decide to re-examine that approach in the future, some of the ground work will already be done.

Reply:
I agree with keeping S3 in our back pocket.

However, I don't understand why memory usage is a problem. The whole point of multipart upload is to never have more than a certain sized chunk of data in flight at once. Just for illustration with made up numbers: rather than holding a 1GB file in memory for validation or uploading or whatever, we'd perform multiple operations on 1MB chunks 1000 times. Settling 1000 1MB chunks into ArangoDB would be a lot of DB connections, but it seems completely limit the memory requirements for the operation.

Reply:
> I agree with keeping S3 in our back pocket.
> 
> However, I don't understand why memory usage is a problem. The whole point of multipart upload is to never have more than a certain sized chunk of data in flight at once. Just for illustration with made up numbers: rather than holding a 1GB file in memory for validation or uploading or whatever, we'd perform multiple operations on 1MB chunks 1000 times. Settling 1000 1MB chunks into ArangoDB would be a lot of DB connections, but it seems completely limit the memory requirements for the operation.

I don't think there's really memory concerns regarding this part of the upload process, just performance concerns really. If we have to chunk a file into 1000 parts, that's a lot of overhead, and thus the upload will be slow. Plus, arango isn't really designed to store binary data. However I don't really know if performance will be an issue in practice, it's just something in my mind.